### ISSUE TITLE
Due to changes in the router, route generation has been temporarily disabled.

### ISSUE THREAD

**Issue by webia1** [angular/angular-cli#1093]
When about it will be available again?

**Comment by liangwang320** [comment]
+1

**Comment by developer-hosny** [comment]
same here

**Comment by filipesilva** [comment]
This message is correct and intended as of beta.6. The motivation for this was as follows:

- `router-deprecated` was, well deprecated, but was being recommended within the community over `router@2.0.0`. 
- `router@2.0.0` was being completely replaced by `router@3.0.0`.
- `router@3.0.0` was introduced in http://angularjs.blogspot.ie/2016/06/improvements-coming-for-routing-in.html with the strong recommendation of `You should begin using this router right away.`.
- `router@3.0.0` is conceptually quite different from any of the two previous routers.
- We didn't have code generation for the new router ready yet.

We felt we wouldn't be doing our users any favors by encouraging the use of `router@2.0.0`, so it would be better to offer no code generation at all.

We're working to re-add router functionality as fast as possible but I don't have a date for it.

**Comment by Brocco** [comment]
To add on to what @filipesilva stated above...

`router@3.0.0` is brand new and the usage of it in regards to the style guide will take time to work itself out. After the style guide has adopted a recommended strategy for using the new router the CLI will evaluate to determine how best to follow the style guide recommendations for route generation. If it is determined that the CLI can add value to that part of the developer workflow we will add those features at that time.

**Comment by ogookafor** [comment]
Waiting...

**Comment by seemsindie** [comment]
+1

**Comment by eudago** [comment]
+1

**Comment by neerajyadav** [comment]
waiting..

**Comment by jasonswett** [comment]
Let me see if I can help clarify something.

@Brocco @filipesilva Based on your comments I get the impression that it may be more on the order of weeks rather than days until we have route generation capabilities again. Do you think this would be about the right expectation? I'm not saying this to complain, I just want to have a feel for what to expect.

**Comment by Brocco** [comment]
@jasonswett I'll try to clarify as best I can.

**We have no idea if route generation will be returning let alone when.**

The best practices for routing from the style guide are being put together right now. Once they are done with that we will take a look and see if the CLI can/should assist with route generation. Routes are disconnected from components now, and we want to ensure that anything that is added to the CLI will be beneficial for the users. Route definitions are rather straightforward IMO now so I would not be surprised if `ng generate route` is deprecated, but that is yet to be determined.

**Comment by kylecordes** [comment]
I have a somewhat different take on this... more predictive in nature:

* It is very likely that Angular 2 will become quite broadly used, with **lots** of tooling available.
* It is very likely that the most popular CLI a year from now, for Angular 2 will offer route management.
* It is very likely that the current Google-blessed cli.angular.io in this repo will grow in to that role.
* But open source is a surprisingly competitive environment. If the current project doesn't get there, it is likely other efforts will pop up. So don't worry too much.
* As to how fast this happens, that is much harder to guess.

**Comment by dancancro** [comment]
@filipesilva & @Brocco in lieu of an automatic generator would you recommend manually following the pattern of [this](http://plnkr.co/edit/?p=preview) demo app from the [documentation](https://angular.io/docs/ts/latest/guide/router.html)? 

What sort of coordination is there between the people making the Angular documentation, the people making the style guide and the people making angular-cli? Are they different people?

**Comment by dancancro** [comment]
> Route definitions are rather straightforward IMO

@Brocco in lieu of a generator could you provide a loose step by step for manually changing an angular-cli app to use the new routing?

Something like

1. Replace system.config.ts with the one from [here](http://plnkr.co/edit/?p=preview) (or add somehow edit package.json)
2. Remove the `@Routes` part of your myapp.component.ts
3. Create a mycomponent.routes.ts file for each component
3. and so on

That would be very helpful. Thanks.

**Comment by seemsindie** [comment]
I am not up to date but wasn't there a command that you all talked about `ng upgrade`?
Why is generation disabled, why it's not left for upgrade to fix the current code?

**Comment by jasonswett** [comment]
This brings up perhaps a good point. I'm not used to the functionality of libraries disappearing even though I haven't upgraded my library version. Wouldn't it be more "normal" to allow `ng generate route` to continue to function in versions <= 1.0.0-beta.6 (or whatever) and deprecate the feature in the next version?

I had written a number of tutorials that depended on `ng route` existing, and now those tutorials are somewhat useless since they're no longer followable. It would be nice if I could just say "this tutorial is valid for Angular CLI <= 1.0.0-beta.6" rather than have to rewrite them and have them be broken in the meantime.

**Comment by ogookafor** [comment]
Exactly! I mean we could have just used "ng route" till it's  deprecation.

On 22 Jun 2016 7:27 p.m., "Jason Swett" <notifications@github.com> wrote:
>
> This brings up perhaps a good point. I'm not used to the functionality of
libraries disappearing even though I haven't upgraded my library version.
Wouldn't it be more "normal" to allow ng generate route to continue to
function in versions <= 1.0.0-beta.6 (or whatever) and deprecate the
feature in the next version?
>
> I had written a number of tutorials that depended on ng route existing,
and now those tutorials are somewhat useless since they're no longer
followable. It would be nice if I could just say "this tutorial is valid
for Angular CLI <= 1.0.0-beta.6" rather than have to rewrite them and have
them be broken in the meantime.

**Comment by ayon06** [comment]
has there been any updates on this?

**Comment by colindevs** [comment]
I understand that the router is evolving quickly, and that the CLI is supposed to uphold best practices, and therefore does not want to lead people down the wrong path by letting people generate routes in a fashion that is likely to be changed soon.

But all I want to know is:
Where can I find some docs on the 3.0.0 router? Or at the 2.0.0.

If we can't automatically generate routes, I'd be happy to manually create them, but I can't seem to find any tutorials. 

Could anybody point me in the right direction?

**Comment by kylecordes** [comment]
The current published angular.io Router docs cover the "3" router.

https://angular.io/docs/ts/latest/guide/router.html

**Comment by colindevs** [comment]
@kylecordes I'm sorry, I assumed that they were wrong because they weren't working in an angular-cli app I am developing. But I just went back and checked, and the problem is that my app is not on the 3 router! My bad, it's fine now...

**Comment by melloradamj** [comment]
Any progress/updates on this?

**Comment by masaanli** [comment]
Waiting for this.

**Comment by val-samonte** [comment]
is there an existing issue reference for the current routing style guide?

**Comment by yusijs** [comment]
@filipesilva @Brocco any chance you can confirm whether or not you'll bring back route generation? Also the tags on this issur suggest its coming ("urgent") but based on @Brocco's comment, there is a solid chance it's not coming at all.

**Comment by Brocco** [comment]
Excellent point @yusijs I just removed that tag, the inclusion of route generation is still being debated/discussed.

**Comment by yusijs** [comment]
@Brocco cool beans, cheers! Looking forward to a solid decision here. Personally I'm fine either way. While I'd prefer route generation being possible, it's hardly a big inconvience adding them manually. I can also see how it can be a bit of a hassle adding it to the cli, since routes can be declared across files, and are no longer coupled with components.

Anyway, what do I know  ¯\_(ツ)_/¯

**Comment by LuisJoseSanchez** [comment]
waiting...

**Comment by delasteve** [comment]
From my PR https://github.com/angular/angular-cli/pull/1113#issuecomment-231451321

@hansl:

> We already have someone internally working on making ng generate route works properly again. The project is limited to route configuration, and does not include guards. I'm going to put this PR on hold until we re-enable route generation and then we can discuss more about adding guards to the project as a new feature.

>Cheers!

**Comment by frankieloscavio** [comment]
waiting... still occurs on Win10 64bit as of today 07/11/16

"Due to changes in the router, route generation has been temporarily disabled. You can find more information about the new router here: http://victorsavkin.com/post/145672529346/angular-router"

**Comment by corasla** [comment]
I am also waiting for this and would have preferred for the warning to exist but let me choose to currently generate the old and deprecated version of the route...

**Comment by yusijs** [comment]
I honestly don't see why people are constantly asking. I'm sure they are aware people are waiting :3

**Comment by webia1** [comment]
@yusijs Thus, it becomes apparent, how much this feature is needed,..

**Comment by yusijs** [comment]
It's hardly needed. Adding a new route takes <5 seconds. Like I mentioned above, it is convenient having it, but considering how easy adding new routes is now, I hardly think it's needed. Desired, yes. :)

**Comment by nraboy** [comment]
Yes it is very easy to implement manually, but I'm not sure I agree with it taking less than five seconds to create a route that way.

Mind sharing your process?

**Comment by yusijs** [comment]
In vs code, I open my routerconfig file, type ng2rcp, hit tab, routerpath, component name. I need to import the component, so maybe 8 seconds)

**Comment by nraboy** [comment]
And the process for creating the files that go with the route configuration? That is where the main convenience of the CLI tool was for me.  Not really in the config as that is quick.

**Comment by yusijs** [comment]
Creating a routerconfig takes slightly longer, but I believe john papas ng2 plugin has router config snippet. So creating the file manually and doing ng2routeconf<tab> isnt a big process.

Dont get me wrong as I'd prefer it in as well. I just don't think it's as important as others do I guess.

**Comment by jffry** [comment]
**tl;dr**: Whatever the ultimate decision on route generation is, I think there is ample room for some documentation improvements in the meantime to make a route-gen-less `angular-cli` a bit less user-hostile.

If you have time and knowledge, please consider contributing PRs to improve the documentation users encounter when attempting to generate routes.

Long version:

For now at least, should the README be update to mention that route generation is disabled, and perhaps also mention why, and what you need to do to make a route manually?

For people who aren't already familiar with setting up routes in an `angular-cli` project, this is the rough journey you might take:

```console
$ npm install --global angular-cli
$ ng init foo-project
$ cd foo-project
$ ng serve
```

"OK that was pretty easy so how do I add a route?  Hrm let's look at the README, Ctrl+F 'route'"

> Generating a route
> You can generate a new route [by running] `ng generate route hero`

"Seems easy enough, let's give it a shot"

```console
$ ng generate route hero
Due to changes in the router, route generation has been temporarily disabled. You can find more information about the new router here: http://victorsavkin.com/post/145672529346/angular-router
```

"OK well that's a little inconvenient, let's go read the article"

> **ANGULAR ROUTER**
> *&lt;snip&gt;2500 words later...&lt;/snip&gt;*
> **E2E Example**

*flips desk*

*throws computer out window*

---

This is, of course, an incredibly frustrating experience.  I'm using `angular-cli` because it's supposed to _help_ me structure an app _the right way_ and yet it provides absolutely no guidance on routing.  If route generation is going to remain disabled for a long time, there's a couple of things that could be done to improve the ergonomics for new-ish users of this tool:

1. The default app scaffolded out by `ng init` should at least demonstrate basic usage of the router, even if `ng generate` won't make new routes.

2. The main README should be updated to mention that route generation is disabled, and ideally include a link to some kind of simple guide of "here's a 101 of how you set up routing in this project you just scaffolded"

I've honestly just started playing with `angular-cli` and it's a major bummer to run into zero guidance on routing.  Maybe I'm just scrabbling away at the bottom of the learning curve and it's really not too bad, but right now `angular-cli` is a tool that's asking me to go along with all its style guide suggestions and then when I go "hey angular-cli what about routes" the answer is "I'm sorry Jeff, I can't do that"

</rant>

Anywho, if you have any examples to help get me up and over the hump, I'd happily consume them and then build a PR to improve the docs to help get other people up over the hump as well.

**Comment by yusijs** [comment]
Those are very valid points. I found the new router easy, but I started off with a.. simpler article at thoughtram: http://blog.thoughtram.io/angular/2016/06/14/routing-in-angular-2-revisited.html

Victors blog post is amazing, but it goes so deep it's not even funny. Pascal's post is a bit easier on my brain. 

Moreover - I agree the readme should be updated - maybe something for @filipesilva or @brocco to look into?

**Comment by hansl** [comment]
So a couple of things that's going on in this issue:

1. Saying "waiting" does nothing to contribute to the discussion. If you have nothing else to add, you can now add emojis to posts so use those to votes on.
2. @jffry you make some good points and I will clarify the README today.
3. we know the amount of frustrations in missing a key component of a CLI, but you don't need the generate action in the CLI for routing to work in your apps. You can generate your own files and the build system will handle it.

We have someone internally working on this, but we want to do it right. We're working on a better framework than just the "string search and replace" we had before (using ASTs instead, for example), and because of it it's taking a bit more time. The route generation will come back in a few weeks, it's not a matter of "if".

Thanks everyone for your patience. Please remember that this is beta and, well, a bit more in flux than what we expected. The change in Router was unfortunate, but in the future we will have a stronger (and faster and better) final release because of this work. So it's a pain for the better, trust me.

Cheers!

**Comment by webia1** [comment]
I make it short: any generator without the ability to create routes is USELESS! 

**A route means at least **

1. Create a directory containing all the necessary files:
2. Add a Component and give it a name (configurable over cli-parameter, route and component may have completely different names)
3. Add an appropriate template file
4. Add an appropriate stylesheet file
5. Add predefined test-spec files (E2E/Unittest/..)
6. (Additional/Optional) Add a Service/Model (configurable over cli-parameter)
7. Last but not least: Register route providers, if it is not already done -> Bootsrapping

You have to do all of them manually, that is too much work!

John Papas great NG2-Snippets unfortunately do not help here,..

**Comment by yusijs** [comment]
1-6 are still done by the cli?

**Comment by filipesilva** [comment]
@jffry @yusijs the official docs also have updated chapters for the new router, and they're being kept up to date with each release:
- https://angular.io/docs/ts/latest/tutorial/toh-pt5.html (bear in mind this is part 5 of the tutorial)
- https://angular.io/docs/ts/latest/guide/router.html

**Comment by yusijs** [comment]
@filipesilva I'm aware, I was answering @webia1 :)

**Comment by masaanli** [comment]
Currently starting with a big AngularJS 2 application (100+ unique pages)

So i need to create 100 routes ;)

What is the approach the Angular CLI is doing?

Put the routes per route-component or a routes.ts file?

**Comment by jasonswett** [comment]
Right now Angular CLI isn't generating routes. I don't exactly how it will implement routes when route generation comes back.

The official Angular docs I've seen seem to want you to put your routes all in one file specifically dedicated to routes.

I've pasted below a route file from an app I'm working on in case it's helpful to see an example.

```typescript
// src/app/app.routes.ts

import { provideRouter, RouterConfig } from '@angular/router';
import { BookListComponent } from './book-list';
import { NewBookComponent } from './new-book';
import { EditBookComponent } from './edit-book';

const routes: RouterConfig = [
  { path: '', component: BookListComponent },
  { path: 'new-book', component: NewBookComponent },
  { path: 'edit-book/:id', component: EditBookComponent }
];

export const appRouterProviders = [
  provideRouter(routes)
];
```

(Any further discussion is probably more appropriately had in some other place.)

**Comment by iBasit** [comment]
Guideline should be put on readme file, for how to use route without generator for temporarily. 

when `route generator` will be available again?

**Comment by radoslavpetranov** [comment]
Any updates on this? Is this feature currently available in master and/or are you planning on enabling it for the next version of CLI that will be switching to RC6? I'm trying to figure out if I should wait some more, if I should just follow the guidelines in your doc and start redoing my deprecated router routes manually or if I should switch to master if rc6 support and route generation are already enabled there.

Shedding some light on this route generation business would be highly appreciated!

**Comment by Brocco** [comment]
Current status on master: adding of routes is still a manual process. But when you generate a module via `ng generate module some-feature` you can specify the option of `routing` via `--routing` which will generating a routing file where you can specify the routes you wish to include within the module.

**Comment by radoslavpetranov** [comment]
Thanks for the update, @Brocco !

**Comment by EasonPai** [comment]
Hi, @Brocco, @radoslavpetranov,
I am 1-week-new to angular, after reading 
`But when you generate a module via ng generate module some-feature you can specify the option of routing via --routing which will generating a routing file where you can specify the routes you wish to include within the module.`

still have no idea how to do routing, I tried this, but no magic, 
`ng generate module some-feature --routing` 

can you give an real example case, thanks

**Comment by webia1** [comment]
@EasonPai 

`ng generate module some-feature --routing `

is correct.  If you change to the directory `some-feature`, you see the files:

```
some-feature-routing.module.ts   
some-feature.component.css       
some-feature.component.html      
some-feature.component.spec.ts   
some-feature.component.ts        
some-feature.module.ts           
```
and in the file `some-feature-routing.module.ts` you will see the routing information:

```
import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

const routes: Routes = [];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
  providers: []
})
export class SomeFeatureRoutingModule { }
```

You see this information in every module, instead of modifying a central file for routing like [in Angular2-Seed from MGechev](https://github.com/mgechev/angular-seed). That's a kind of "using a sledge-hammer to crack a nut",..

I wonder, why not learning from one another??

**Comment by Meligy** [comment]
@Brocco can we please remove the route generator completely? Or change the message to say use `ng generate route --module` instead?

Cheers,

**Comment by Brocco** [comment]
Current status of route generation:

You can generate routing modules on new app creation and on module generation using the flag `--routing`. As of now, generation of individual routes will not be supported.

**Comment by Meligy** [comment]
@Brocco we probably should change this message into something that explains this instead.

Something along the lines of:

> To generate a route, use `ng generate module` with the ` --routing` flag
> You can then add your individual routes to the moduleName-routing.module.ts file

Happy to send a PR if the copy is OK, or to use alternative suggested copy.

Cheers,

**Comment by Brocco** [comment]
@Meligy A PR with that update would be great!

**Comment by peterpeterparker** [comment]
@Brocco shouldn't be the online documentation updated about that?

https://cli.angular.io/reference.pdf

**Comment by Meligy** [comment]
- Interestingly, the `route` blueprint has been deleted (although the alias `r` is not cleaned up).
- The docs in readme need updating.
- The PDF is very outdated. It'd make sense for cli.angular.io to link to the readme instead.

**Comment by RaphHaddad** [comment]
any status on this PR? @Meligy ?

**Comment by Meligy** [comment]
@RaphHaddad I have updated the README in #3983 and the PR has been approved, so it should be merged soon.

For now, you can review the updated docs here: https://github.com/Meligy/angular-cli/blob/b391ef8811b0e6c0a3b549708764f814bf860d12/README.md#generating-a-route

**Comment by Meligy** [comment]
@RaphHaddad actually it's already cherry picked into master, you can review the newest README from the official repository here https://github.com/angular/angular-cli#generating-a-route

**Comment by ValerianMenin** [comment]
Should we use [](https://angular.io/docs/ts/latest/guide/router.html) or [](https://github.com/angular/angular-cli#generating-a-route)
I'm a bit confused on how it works. I'm learning how to use Angular2 and I find so many different ways... Which one is the lastest or best practice ?

**Comment by danday74** [comment]
please give us some idea of a release date - thanks

**Comment by yoneal** [comment]
@ValerianMenin If I'm not mistaken, cli is using the ff https://angular.io/docs/ts/latest/guide/router.html#!#routing-module

**Comment by Brocco** [comment]
Specific route generation is not on the roadmap going forward, you can get started with routing by specifying `--routing` on new project generations and when generating modules. This will provide a routing module to be used to define your routes.

I am locking this thread, if you have issues with what is currently being generated please open a new issue.
